### PR TITLE
KAFKA-12578: Remove deprecated security classes/methods for 3.0

### PR DESCRIPTION
**Pull request by ijuma**
More specifically, remove deprecated:
- Constants in SslConfigs
- Constants in SaslConfigs
- AclBinding constructor
- AclBindingFilter constructor
- PrincipalBuilder and DefaultPrincipalBuilder classes
- ResourceFilter

Also simplify tests and code that no longer have to handle the removed `PrincipalBuilder`.

These removals seem non controversial. There is a straightforward alternative. The
deprecations happened in 1.0.0 and 2.0.0.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by ijuma** [comment]
I still need to update the upgrade notes.

File: clients/src/main/java/org/apache/kafka/common/security/authenticator/DefaultKafkaPrincipalBuilder.java [hunk 1]
```diff
@@ -41,89 +39,39 @@
...
     /**
      * Construct a new instance.
      *
      * @param kerberosShortNamer Kerberos name rewrite rules or null if none have been configured
      * @param sslPrincipalMapper SSL Principal mapper or null if none have been configured
      */
     public DefaultKafkaPrincipalBuilder(KerberosShortNamer kerberosShortNamer, SslPrincipalMapper sslPrincipalMapper) {
-        this(null, null, null, kerberosShortNamer, sslPrincipalMapper);
+        this.kerberosShortNamer = kerberosShortNamer;
+        this.sslPrincipalMapper =  sslPrincipalMapper;
```
Comment (chia7712) [inline clients/src/main/java/org/apache/kafka/common/security/authenticator/DefaultKafkaPrincipalBuilder.java]: redundant "space"

File: clients/src/main/java/org/apache/kafka/common/acl/AclBindingFilter.java [hunk 2]
```diff
@@ -50,18 +48,6 @@ public AclBindingFilter(ResourcePatternFilter patternFilter, AccessControlEntryF
...
 
-    /**
-     * Create an instance of this filter with the provided parameters.
-     *
-     * @param resourceFilter non-null resource filter
-     * @param entryFilter non-null access control entry filter
-     * @deprecated Since 2.0. Use {@link #AclBindingFilter(ResourcePatternFilter, AccessControlEntryFilter)}
-     */
-    @Deprecated
-    public AclBindingFilter(ResourceFilter resourceFilter, AccessControlEntryFilter entryFilter) {
```
Comment (chia7712) [inline clients/src/main/java/org/apache/kafka/common/acl/AclBindingFilter.java:61]: Is `ResourceFilter` still useful for production?

**Review by chia7712 — APPROVED** [review]
@ijuma Nice cleanup. LGTM

Could you update the docs (https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/common/config/internals/BrokerSecurityConfigs.java#L42)?

File: clients/src/main/java/org/apache/kafka/common/acl/AclBindingFilter.java (hunk 2 above)
Comment (ijuma) [reply clients/src/main/java/org/apache/kafka/common/acl/AclBindingFilter.java:61]: Good point, this is not used anymore. It's not deprecated, but given that it's not used anywhere, it seems fair to remove it at the same time. Thoughts?

Comment (chia7712) [reply clients/src/main/java/org/apache/kafka/common/acl/AclBindingFilter.java:61]: +1 to remove it :)

**Comment by ijuma** [comment]
Updated @chia7712

**Comment by chia7712** [comment]
Is `Resource` exposed to users? It appears that we don't create `Resource` except for `Resource CLUSTER`. If `Resource` is not a kind of public APIs, I feel it can be removed also. (`Resource.CLUSTER` can be replaced by `ResourceType` and `String`)

**Comment by ijuma** [comment]
@chia7712 `Resource` is public API, yes. But you're right that it may not be used via client facing APIs. Do you mind if we tackle this one separately?

**Comment by chia7712** [comment]
> Do you mind if we tackle this one separately?

+1

**Pull request merged**